pubsub/awssnssqs/awssnssqs.go

```diff
@@ -553,28 +553,69 @@ func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
 		}
 		return nil
 	}
+	return s.handleAcksInBatches(ctx, ids, newReq, addToReq, sendReq)
+}
+
+// SendNacks implements driver.Subscription.SendNacks.
+func (s *subscription) SendNacks(ctx context.Context, ids []driver.AckID) error {
+	newReq := func() interface{} {
+		return &sqs.ChangeMessageVisibilityBatchInput{QueueUrl: aws.String(s.qURL)}
+	}
+	addToReq := func(reqi interface{}, id driver.AckID) {
+		req := reqi.(*sqs.ChangeMessageVisibilityBatchInput)
+		req.Entries = append(req.Entries, &sqs.ChangeMessageVisibilityBatchRequestEntry{
+			Id:                aws.String(strconv.Itoa(len(req.Entries))),
+			ReceiptHandle:     id.(*string),
+			VisibilityTimeout: aws.Int64(0),
+		})
+	}
+	sendReq := func(ctx context.Context, reqi interface{}) error {
+		req := reqi.(*sqs.ChangeMessageVisibilityBatchInput)
+		resp, err := s.client.ChangeMessageVisibilityBatchWithContext(ctx, req)
+		if err != nil {
+			return err
+		}
+		if numFailed := len(resp.Failed); numFailed > 0 {
+			first := resp.Failed[0]
+			return awserr.New(aws.StringValue(first.Code), fmt.Sprintf("sqs.ChangeMessageVisibilityBatch failed for %d message(s): %s", numFailed, aws.StringValue(first.Message)), nil)
+		}
+		return nil
+	}
+	return s.handleAcksInBatches(ctx, ids, newReq, addToReq, sendReq)
+}
+
+func (s *subscription) handleAcksInBatches(ctx context.Context, ids []driver.AckID, newReq func() interface{}, addToReq func(req interface{}, id driver.AckID), sendReq func(ctx context.Context, req interface{}) error) error {
+	// SQS supports deleting/updating at most 10 messages at a time:
+	// https://godoc.org/github.com/aws/aws-sdk-go/service/sqs#SQS.DeleteMessageBatch
+	// https://godoc.org/github.com/aws/aws-sdk-go/service/sqs#SQS.ChangeMessageVisibilityBatch
+	// If len(ids) is larger than 10, we make up to maxAckConcurrency RPCs
+	// (across calls to SendAcks/SendNacks) in parallel.
+	const maxPerRPC = 10
 
-	req := &sqs.DeleteMessageBatchInput{QueueUrl: aws.String(s.qURL)}
+	g, ctx := errgroup.WithContext(ctx)
+	var ctxErr error
+
+	req := newReq()
+	n := 0
 Loop:
 	for i, id := range ids {
-		req.Entries = append(req.Entries, &sqs.DeleteMessageBatchRequestEntry{
-			Id:            aws.String(strconv.Itoa(i)),
-			ReceiptHandle: id.(*string),
-		})
+		addToReq(req, id)
+		n++
 		// Send if we've reached the limit, or if this is the last one.
-		if len(req.Entries) == maxPerRPC || i == len(ids)-1 {
+		if n == maxPerRPC || i == len(ids)-1 {
 			select {
 			case <-ctx.Done():
 				ctxErr = ctx.Err()
 				break Loop
 			case s.acksem <- struct{}{}:
 				req := req
 				g.Go(func() error {
-					return sendBatch(req)
+					defer func() { <-s.acksem }()
+					return sendReq(ctx, req)
 				})
 			}
 			// Start the next batch.
-			req = &sqs.DeleteMessageBatchInput{QueueUrl: aws.String(s.qURL)}
+			req = newReq()
```

Comment:
Set `n = 0` around here, I would think.

If I'm right, your tests aren't catching this, so fix that too.

Reply:
Good catch. I think this is hard to test today, because any test that does more than 10 acks is going to be flaky due to the ack batcher. I could write an AWS-specific test of this function, but I'd rather write a conformance test; I'll do so as part of #1698.

Reply:
(fixed)